From ca28b8b125c27063b9b4bc60bb85206ca8e0d403 Mon Sep 17 00:00:00 2001
From: Yuvasree Sivasankaran <quic_ysivasan@quicinc.com>
Date: Thu, 31 Aug 2023 10:59:33 +0530
Subject: [PATCH] wifi: mac80211: Add mac hw flag to avoid queue skb

Queue SKB in mac80211 become mandatory from latest 6.1 kernel. Because of
this queuing, there will be performance degradation. Add hw flag option
to enable tx queue in Driver/Hardware.

Driver/hardware can register for HAS_TX_QUEUE HW flag and avoid tx queuing
in mac80211.

Add same HW flag checks to avoid accessing skb queues which will be
NULL or invalid and also NULL checks for sta txqs for NULL or invalid
access.

Signed-off-by: Yuvasree Sivasankaran <quic_ysivasan@quicinc.com>
---
 include/net/mac80211.h     |    4 ++
 net/mac80211/debugfs.c     |    1 +
 net/mac80211/tx.c          |   23 +++++++++----
 net/mac80211/cfg.c         |    3 ++
 net/mac80211/rx.c          |    3 ++
 net/mac80211/sta_info.c    |   75 ++++++++++++++++++++++++++------------------
 net/mac80211/debugfs_sta.c |    4 ++
 net/mac80211/mesh.c        |    3 +-
 net/mac80211/util.c        |    3 ++
 net/mac80211/ht.c          |   54 +++++++++++++++++---------------
 10 files changed, 109 insertions(+), 64 deletions(-)

--- a/include/net/mac80211.h
+++ b/include/net/mac80211.h
@@ -3125,6 +3125,9 @@ struct ieee80211_txq {
  *
  * @IEEE80211_HW_SUPPORTS_TID_CLASS_OFFLOAD: Hardware suports tid calssification offload.
  *
+ * @IEEE80211_HW_HAS_TX_QUEUE: Hardware/drivers has tx queue, does skb queuing itself,
+ *	the stack will not do tx queuing.
+ *
  * @NUM_IEEE80211_HW_FLAGS: number of hardware flags, used for sizing arrays
  */
 enum ieee80211_hw_flags {
@@ -3189,6 +3192,7 @@ enum ieee80211_hw_flags {
 	IEEE80211_HW_SUPPORTS_NSS_OFFLOAD,
 	IEEE80211_HW_SUPPORTS_MESH_NSS_OFFLOAD,
 	IEEE80211_HW_SUPPORTS_TID_CLASS_OFFLOAD,
+	IEEE80211_HW_HAS_TX_QUEUE,
 
 	/* keep last, obviously */
 	NUM_IEEE80211_HW_FLAGS
--- a/net/mac80211/debugfs.c
+++ b/net/mac80211/debugfs.c
@@ -486,6 +486,7 @@ static const char *hw_flag_names[] = {
 	FLAG(SUPPORTS_NSS_OFFLOAD),
 	FLAG(SUPPORTS_MESH_NSS_OFFLOAD),
 	FLAG(SUPPORTS_TID_CLASS_OFFLOAD),
+	FLAG(HAS_TX_QUEUE),
 #undef FLAG
 };
 
--- a/net/mac80211/tx.c
+++ b/net/mac80211/tx.c
@@ -1679,7 +1679,8 @@ static bool ieee80211_queue_skb(struct i
 	struct ieee80211_vif *vif;
 	struct txq_info *txqi;
 
-	if (sdata->vif.type == NL80211_IFTYPE_MONITOR)
+	if (ieee80211_hw_check(&local->hw, HAS_TX_QUEUE) ||
+	    sdata->vif.type == NL80211_IFTYPE_MONITOR)
 		return false;
 
 	if (sdata->vif.type == NL80211_IFTYPE_AP_VLAN)
@@ -4445,7 +4446,8 @@ void __ieee80211_subif_start_xmit(struct
 		}
 	}
 
-	skb_set_queue_mapping(skb, ieee80211_select_queue(sdata, sta, skb));
+	if (unlikely(!ieee80211_hw_check(&local->hw, HAS_TX_QUEUE)))
+		skb_set_queue_mapping(skb, ieee80211_select_queue(sdata, sta, skb));
 	ieee80211_aggr_check(sdata, sta, skb);
 
 	if (sta) {
@@ -4801,8 +4803,12 @@ static void ieee80211_8023_xmit(struct i
 	bool multicast;
 	u8 tid;
 
-	queue = ieee80211_select_queue(sdata, sta, skb);
-	skb_set_queue_mapping(skb, queue);
+	queue = skb_get_queue_mapping(skb);
+
+	if (unlikely(!ieee80211_hw_check(&local->hw, HAS_TX_QUEUE))) {
+		queue = ieee80211_select_queue(sdata, sta, skb);
+		skb_set_queue_mapping(skb, queue);
+	}
 
 	multicast = is_multicast_ether_addr(ra);
 
@@ -6799,9 +6805,12 @@ int ieee80211_tx_control_port(struct wip
 	}
 
 	if (!IS_ERR(sta)) {
-		u16 queue = ieee80211_select_queue(sdata, sta, skb);
 
-		skb_set_queue_mapping(skb, queue);
+		if (!ieee80211_hw_check(&local->hw, HAS_TX_QUEUE)) {
+			u16 queue = ieee80211_select_queue(sdata, sta, skb);
+
+			skb_set_queue_mapping(skb, queue);
+		}
 
 		/*
 		 * for MLO STA, the SA should be the AP MLD address, but
--- a/net/mac80211/cfg.c
+++ b/net/mac80211/cfg.c
@@ -5414,6 +5414,9 @@ static int ieee80211_get_txq_stats(struc
 	struct ieee80211_sub_if_data *sdata;
 	int ret = 0;
 
+	if (ieee80211_hw_check(&local->hw, HAS_TX_QUEUE))
+		return 1;
+
 	spin_lock_bh(&local->fq.lock);
 
 	if (wdev) {
--- a/net/mac80211/rx.c
+++ b/net/mac80211/rx.c
@@ -1703,6 +1703,9 @@ static void sta_ps_start(struct sta_info
 
 	ieee80211_clear_fast_xmit(sta);
 
+	if (!sta->sta.txq[0])
+		return;
+
 	for (tid = 0; tid < IEEE80211_NUM_TIDS; tid++) {
 		struct ieee80211_txq *txq = sta->sta.txq[tid];
 		struct txq_info *txqi = to_txq_info(txq);
--- a/net/mac80211/sta_info.c
+++ b/net/mac80211/sta_info.c
@@ -117,15 +117,17 @@ void ieee80211_purge_sta_txqs(struct sta
 	struct ieee80211_local *local = sta->sdata->local;
 	int i;
 
-	for (i = 0; i < ARRAY_SIZE(sta->sta.txq); i++) {
-		struct txq_info *txqi;
+	if (sta->sta.txq[0]) {
+		for (i = 0; i < ARRAY_SIZE(sta->sta.txq); i++) {
+			struct txq_info *txqi;
 
-		if (!sta->sta.txq[i])
-			continue;
+			if (!sta->sta.txq[i])
+				continue;
 
-		txqi = to_txq_info(sta->sta.txq[i]);
+			txqi = to_txq_info(sta->sta.txq[i]);
 
-		ieee80211_txq_purge(local, txqi);
+			ieee80211_txq_purge(local, txqi);
+		}
 	}
 }
 
@@ -502,7 +504,9 @@ void sta_info_free(struct ieee80211_loca
 
 	sta_dbg(sta->sdata, "Destroyed STA %pM\n", sta->sta.addr);
 
-	kfree(to_txq_info(sta->sta.txq[0]));
+	if (sta->sta.txq[0])
+		kfree(to_txq_info(sta->sta.txq[0]));
+
 	kfree(rcu_dereference_raw(sta->sta.rates));
 #ifdef CPTCFG_MAC80211_MESH
 	kfree(sta->mesh);
@@ -637,8 +641,6 @@ __sta_info_alloc(struct ieee80211_sub_if
 	struct ieee80211_local *local = sdata->local;
 	struct ieee80211_hw *hw = &local->hw;
 	struct sta_info *sta;
-	void *txq_data;
-	int size;
 	int i;
 
 	sta = kzalloc(sizeof(*sta) + hw->sta_data_size, gfp);
@@ -711,18 +713,22 @@ __sta_info_alloc(struct ieee80211_sub_if
 
 	sta->last_connected = ktime_get_seconds();
 
-	size = sizeof(struct txq_info) +
-	       ALIGN(hw->txq_data_size, sizeof(void *));
 
-	txq_data = kcalloc(ARRAY_SIZE(sta->sta.txq), size, gfp);
-	if (!txq_data)
-		goto free;
+	if (!ieee80211_hw_check(&local->hw, HAS_TX_QUEUE)) {
+		void *txq_data;
+		int size = sizeof(struct txq_info) +
+			ALIGN(hw->txq_data_size, sizeof(void *));
+
+		txq_data = kcalloc(ARRAY_SIZE(sta->sta.txq), size, gfp);
+		if (!txq_data)
+			goto free;
 
-	for (i = 0; i < ARRAY_SIZE(sta->sta.txq); i++) {
-		struct txq_info *txq = txq_data + i * size;
+		for (i = 0; i < ARRAY_SIZE(sta->sta.txq); i++) {
+			struct txq_info *txq = txq_data + i * size;
 
-		/* might not do anything for the (bufferable) MMPDU TXQ */
-		ieee80211_txq_init(sdata, sta, txq, i);
+			/* might not do anything for the (bufferable) MMPDU TXQ */
+			ieee80211_txq_init(sdata, sta, txq, i);
+		}
 	}
 
 	if (sta_prepare_rate_control(local, sta, gfp))
@@ -791,7 +797,8 @@ __sta_info_alloc(struct ieee80211_sub_if
 	return sta;
 
 free_txq:
-	kfree(to_txq_info(sta->sta.txq[0]));
+	if (sta->sta.txq[0])
+		kfree(to_txq_info(sta->sta.txq[0]));
 free:
 	sta_info_free_link(&sta->deflink);
 #ifdef CPTCFG_MAC80211_MESH
@@ -1119,6 +1126,9 @@ static void __sta_info_recalc_tim(struct
 	int ac;
 	u16 id = sta->sta.aid;
 
+	if (ieee80211_hw_check(&local->hw, HAS_TX_QUEUE))
+		return;
+
 	if (sta->sdata->vif.type == NL80211_IFTYPE_AP ||
 	    sta->sdata->vif.type == NL80211_IFTYPE_AP_VLAN) {
 		if (WARN_ON_ONCE(!sta->sdata->bss))
@@ -1859,11 +1869,13 @@ void ieee80211_sta_ps_deliver_wakeup(str
 	if (!ieee80211_hw_check(&local->hw, AP_LINK_PS))
 		drv_sta_notify(local, sdata, STA_NOTIFY_AWAKE, &sta->sta);
 
-	for (i = 0; i < ARRAY_SIZE(sta->sta.txq); i++) {
-		if (!sta->sta.txq[i] || !txq_has_queue(sta->sta.txq[i]))
-			continue;
+	if (!ieee80211_hw_check(&local->hw, HAS_TX_QUEUE)) {
+		for (i = 0; i < ARRAY_SIZE(sta->sta.txq); i++) {
+			if (!sta->sta.txq[i] || !txq_has_queue(sta->sta.txq[i]))
+				continue;
 
-		schedule_and_wake_txq(local, to_txq_info(sta->sta.txq[i]));
+			schedule_and_wake_txq(local, to_txq_info(sta->sta.txq[i]));
+		}
 	}
 
 	skb_queue_head_init(&pending);
@@ -2750,7 +2762,8 @@ static void sta_set_tidstats(struct sta_
 			link_sta_info->status_stats.msdu_failed[tid];
 	}
 
-	if (link_id < 0 && tid < IEEE80211_NUM_TIDS) {
+	if (link_id < 0 && tid < IEEE80211_NUM_TIDS &&
+	    !ieee80211_hw_check(&local->hw, HAS_TX_QUEUE)) {
 		spin_lock_bh(&local->fq.lock);
 
 		tidstats->filled |= BIT(NL80211_TID_STATS_TXQ_STATS);
--- a/net/mac80211/debugfs_sta.c
+++ b/net/mac80211/debugfs_sta.c
@@ -155,6 +155,9 @@ static ssize_t sta_aqm_read(struct file
 		       bufsz + buf - p,
 		       "tid ac backlog-bytes backlog-packets new-flows drops marks overlimit collisions tx-bytes tx-packets flags\n");
 
+	if (ieee80211_hw_check(&local->hw, HAS_TX_QUEUE))
+		goto skip_txq_info;
+
 	for (i = 0; i < ARRAY_SIZE(sta->sta.txq); i++) {
 		if (!sta->sta.txq[i])
 			continue;
@@ -179,6 +182,7 @@ static ssize_t sta_aqm_read(struct file
 			       test_bit(IEEE80211_TXQ_DIRTY, &txqi->flags) ? " DIRTY" : "");
 	}
 
+skip_txq_info:
 	spin_unlock_bh(&local->fq.lock);
 
 	rv = simple_read_from_buffer(userbuf, count, ppos, buf, p - buf);
--- a/net/mac80211/mesh.c
+++ b/net/mac80211/mesh.c
@@ -802,7 +802,8 @@ bool ieee80211_mesh_xmit_fast(struct iee
 	if (!skb)
 		return true;
 
-	skb_set_queue_mapping(skb, ieee80211_select_queue(sdata, sta, skb));
+	if (unlikely(!ieee80211_hw_check(&sdata->local->hw, HAS_TX_QUEUE)))
+		skb_set_queue_mapping(skb, ieee80211_select_queue(sdata, sta, skb));
 
 	meshhdr = (struct ieee80211s_hdr *)entry->hdr;
 	if ((meshhdr->flags & MESH_FLAGS_AE) == MESH_FLAGS_AE_A5_A6) {
--- a/net/mac80211/util.c
+++ b/net/mac80211/util.c
@@ -485,6 +485,9 @@ static void __ieee80211_wake_queue(struc
 	 * release someone's lock, but it is fine because all the callers of
 	 * __ieee80211_wake_queue call it right before releasing the lock.
 	 */
+	if (ieee80211_hw_check(&local->hw, HAS_TX_QUEUE))
+		return;
+
 	if (reason == IEEE80211_QUEUE_STOP_REASON_DRIVER)
 		tasklet_schedule(&local->wake_txqs_tasklet);
 	else
--- a/net/mac80211/ht.c
+++ b/net/mac80211/ht.c
@@ -362,35 +362,39 @@ void ieee80211_ba_session_work(struct wi
 
 		tid_tx = sta->ampdu_mlme.tid_start_tx[tid];
 		if (!blocked && tid_tx) {
-			struct txq_info *txqi = to_txq_info(sta->sta.txq[tid]);
-			struct ieee80211_sub_if_data *sdata =
-				vif_to_sdata(txqi->txq.vif);
-			struct fq *fq = &sdata->local->fq;
-
-			spin_lock_bh(&fq->lock);
-
-			/* Allow only frags to be dequeued */
-			set_bit(IEEE80211_TXQ_STOP, &txqi->flags);
-
-			if (!skb_queue_empty(&txqi->frags)) {
-				/* Fragmented Tx is ongoing, wait for it to
-				 * finish. Reschedule worker to retry later.
-				 */
+			if (sta->sta.txq[tid]) {
+				struct txq_info *txqi =
+					to_txq_info(sta->sta.txq[tid]);
+				struct ieee80211_sub_if_data *sdata =
+					vif_to_sdata(txqi->txq.vif);
+				struct fq *fq = &sdata->local->fq;
+
+				spin_lock_bh(&fq->lock);
+
+				/* Allow only frags to be dequeued */
+				set_bit(IEEE80211_TXQ_STOP, &txqi->flags);
+
+				if (!skb_queue_empty(&txqi->frags)) {
+					/* Fragmented Tx is ongoing, wait for it to
+					 * finish. Reschedule worker to retry later.
+					 */
+
+					spin_unlock_bh(&fq->lock);
+					spin_unlock_bh(&sta->lock);
+
+					/* Give the task working on the txq a chance
+					 * to send out the queued frags
+					 */
+					synchronize_net();
+
+					wiphy_work_queue(sdata->local->hw.wiphy,
+							 work);
+					return;
+				}
 
 				spin_unlock_bh(&fq->lock);
-				spin_unlock_bh(&sta->lock);
-
-				/* Give the task working on the txq a chance
-				 * to send out the queued frags
-				 */
-				synchronize_net();
-
-				wiphy_work_queue(sdata->local->hw.wiphy, work);
-				return;
 			}
 
-			spin_unlock_bh(&fq->lock);
-
 			/*
 			 * Assign it over to the normal tid_tx array
 			 * where it "goes live".
